### ISSUE TITLE
[General Usage]: Why do I need to fully qualify snowflake_grant_privileges_to_account_role.account_role_name?

### ISSUE THREAD

**Issue by jamiekt** [Snowflake-Labs/terraform-provider-snowflake#2982]
### Terraform CLI Version

1.9.3

### Terraform Provider Version

0.94.1

### Terraform Configuration

I have some Snowflake role grants that I want to bring under the control of terraform, so I need to import them. Here is my code:

```hcl
locals {
   monitor_usage_task_roles = toset([
    data.snowflake_role.MY_ROLE1.name,
    data.snowflake_role.MY_ROLE2.name,
  ])
}
import {
  for_each = local.monitor_usage_task_roles
  to       = snowflake_grant_privileges_to_account_role.monitor_usage_task[each.key]
  id       = "${each.key}|false|false|MONITOR USAGE|OnAccount"
}
resource "snowflake_grant_privileges_to_account_role" "monitor_usage_task" {
  for_each          = local.monitor_usage_task_roles
  privileges        = ["MONITOR USAGE"]
  account_role_name = each.key
  with_grant_option = false
  on_account        = true
}
```

The resulting plan looks like this for all those roles:

```
    # snowflake_grant_privileges_to_account_role.monitor_usage_task["MY_ROLE1"] must be replaced
    # (imported from "MY_ROLE1|false|false|MONITOR USAGE|OnAccount")
    # Warning: this will destroy the imported resource
  -/+ resource "snowflake_grant_privileges_to_account_role" "monitor_usage_task" {
        ~ account_role_name = "\"MY_ROLE1\"" -> "MY_ROLE1" # forces replacement
          all_privileges    = false
          always_apply      = false
        ~ id                = "MY_ROLE1|false|false|MONITOR USAGE|OnAccount" -> (known after apply)
          on_account        = true
          privileges        = [
              "MONITOR USAGE",
          ]
          with_grant_option = false
      }
```

Obviously `# Warning: this will destroy the imported resource` is very worrying, I definitely don't want to do that. 

I note from https://registry.terraform.io/providers/Snowflake-Labs/snowflake/latest/docs/resources/grant_privileges_to_account_role#import that

> All the …_name parts should be fully qualified names (where every part is quoted)

So I change my import block to this (note the new escaped quotes in the id of the import):

```hcl
import {
  for_each = local.monitor_usage_task_roles
  to       = snowflake_grant_privileges_to_account_role.monitor_usage_task[each.key]
  id       = "\"${each.key}\"|false|false|MONITOR USAGE|OnAccount"
}
```

The new plan is:

```
    # snowflake_grant_privileges_to_account_role.monitor_usage_task["MY_ROLE1"] must be replaced
    # (imported from ""MY_ROLE1"|false|false|MONITOR USAGE|OnAccount")
    # Warning: this will destroy the imported resource
  -/+ resource "snowflake_grant_privileges_to_account_role" "monitor_usage_task" {
        ~ account_role_name = "\"MY_ROLE1\"" -> "MY_ROLE1" # forces replacement
          all_privileges    = false
          always_apply      = false
        ~ id                = "\"MY_ROLE1\"|false|false|MONITOR USAGE|OnAccount" -> (known after apply)
          on_account        = true
          privileges        = [
              "MONITOR USAGE",
          ]
          with_grant_option = false
      }
```
Note the warning, my resource is still going to be destroyed due to the altered `account_role_name`

Hence I add the same escaping to `account_role_name`

```hcl
resource "snowflake_grant_privileges_to_account_role" "monitor_usage_task" {
  for_each          = local.monitor_usage_task_roles
  privileges        = ["MONITOR USAGE"]
  account_role_name = "\"${each.key}\""
  with_grant_option = false
  on_account        = true
}
```

This time my plan looks better

```
    # snowflake_grant_privileges_to_account_role.monitor_usage_task["MY_ROLE1"] will be imported
      resource "snowflake_grant_privileges_to_account_role" "monitor_usage_task" {
          account_role_name = "\"MY_ROLE1\""
          all_privileges    = false
          always_apply      = false
          id                = "\"MY_ROLE1\"|false|false|MONITOR USAGE|OnAccount"
          on_account        = true
          privileges        = [
              "MONITOR USAGE",
          ]
          with_grant_option = false
      }
```

This doesn't *feel* right though. If I was defining brand new privileges for a role I wouldn't have to wrap `account_role_name` with escaped quotes, so why do I have to do it when importing? That import block will be removed after the objects are imported so I'm left with escaped quotes around `account_role_name` with no real reason for them being there (other than they were required for importing). Moreover, I'm concerned that the role refered to when this gets applied will be
* `"MY_ROLE1"`
 as opposed to
* `MY_ROLE1`

This seems like a strange stipulation. What is the peculiar circumstance that requires me to use escaped quotes when importing?

### Category

category:resource

### Object type(s)

resource:grant_privileges_to_account_role

### Expected Behavior

I would have expected this code to be sufficient

```hcl
resource "snowflake_grant_privileges_to_account_role" "monitor_usage_task" {
  for_each          = local.monitor_usage_task_roles
  privileges        = ["MONITOR USAGE"]
  account_role_name = each.key
  with_grant_option = false
  on_account        = true
}
```

### Actual Behavior

I'm required to use

```hcl
resource "snowflake_grant_privileges_to_account_role" "monitor_usage_task" {
  for_each          = local.monitor_usage_task_roles
  privileges        = ["MONITOR USAGE"]
  account_role_name = "\"${each.key}\""
  with_grant_option = false
  on_account        = true
}
```

### Steps to Reproduce

Create a role in the Snowflake UI
Use the code provided above to try to import it

### How much impact is this issue causing?

Low

### Additional Information

I don't think so, but if other info is required please let me know

**Comment by sfc-gh-jcieslak** [comment]
Hey @jamiekt 👋 
Thank you for reporting the issue. As I'm currently working on identifiers, I'll take a look at it as soon as possible and try to fix it. At first glance at your config and the code we have right now, I don't see any reason why this
```terraform
import {
  for_each = local.monitor_usage_task_roles
  to       = snowflake_grant_privileges_to_account_role.monitor_usage_task[each.key]
  id       = "\"${each.key}\"|false|false|MONITOR USAGE|OnAccount"
}
```
wouldn't work. After I take care of other pending cases, I'll take a closer look at this, but it seems like a typical case of quote mismatch case that we have in some of the other resources we're currently fixing.

**Comment by jamiekt** [comment]
Hi @sfc-gh-jcieslak , thank you very much that is much appreciated.

**Comment by jamiekt** [comment]
Hi @sfc-gh-jcieslak , 
Are you able to give an update on this issue?

Thanks in advance.

**Comment by sfc-gh-jcieslak** [comment]
Yeah, I'm currently working on it, and after that change the quoting should not matter. It'll most likely be a part of the next release (somewhere around next week).

**Comment by jamiekt** [comment]
@sfc-gh-jcieslak Awesome, thank you so much.

**Comment by jamiekt** [comment]
@sfc-gh-jcieslak  When you release this fix will you also do a patch release for 0.92.X? The reason for my asking is that in my PR which is currently blocked by this I am using removed blocks to remove deprecated resources such as `snowflake_account_grant`, e.g.:
```terraform
removed {
  from = snowflake_account_grant.monitor_task_execution
  lifecycle {
    destroy = false
  }
}
```
For that reason I need to use provider version `"<0.93.0"`. 
However, in the same PR I'm also importing the objects created by the resources that are being removed. 
```terraform
import {
  for_each = local.monitor_task_execution_roles
  to       = snowflake_grant_privileges_to_account_role.monitor_task_execution[each.key]
  id       = "\"${each.key}\"|false|false|MONITOR EXECUTION|OnAccount"
}
```
and this is what is causing the problem that we've discussed above.

Hence I need a new version of the provider with the fix you're working on now but which is still `"<0.93.0"`. Hope that makes sense.

**Comment by sfc-gh-jcieslak** [comment]
Hey
I'm sorry, but currently, as we are 0.X.0 we're not supporting backward bugfixes. Even If we would like to do it, our release process is pretty limited at the moment, and it would take us some time to release such a version and make sure that the next versions also have that bugfix. Please, for now, use the quoted version. Starting from the v0.95.0 quoting shouldn't matter.

**Comment by sfc-gh-jcieslak** [comment]
Hey 👋
Yesterday, we released a new version of the provider ([v0.95.0](https://github.com/Snowflake-Labs/terraform-provider-snowflake/releases/tag/v0.95.0)) that shouldn't produce any plans related to quoting. We also prepared a document concluding identifiers rework ([here](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/docs/technical-documentation/identifiers_rework_design_decisions.md)) that should clear things up regarding identifiers.

**Comment by jamiekt** [comment]
Great, thanks @sfc-gh-jcieslak . I'll close this issue and re-open if I find any issues in v.0.95.0

**Comment by jamiekt** [comment]
Hi @sfc-gh-jcieslak ,
Hoping I can get your advice on something. We are currently replacing the resource types that have been removed in 0.93. How we're doing that is:

* pinning to 0.92.0
* replacing the old resource types with their replacement counterparts
* using `removed` blocks to remove the replaced resources from the state file
* importing the definition of the new resource type

There are a lot of resources that we're having to do this for.

The aim here is to produce a plan with zero changes but the plan currently looks like this:

> Plan: 231 to import, 206 to add, 0 to change, 205 to destroy.

and the reason is this quoted identifier problem that we discussed above. We have 205 resources that appear in the plan like this:

```
  # module.resources.module.warehouse[0].snowflake_grant_privileges_to_account_role.warehouse_usage_grant["redacted"] must be replaced
  # (imported from "ROLE_NAME|false|false|USAGE|OnAccountObject|WAREHOUSE|WAREHOUSE_NAME")
  # Warning: this will destroy the imported resource
-/+ resource "snowflake_grant_privileges_to_account_role" "warehouse_usage_grant" {
      ~ account_role_name = "\"ROLE_NAME\"" -> "ROLE_NAME" # forces replacement
        all_privileges    = false
        always_apply      = false
      ~ id                = "ROLE_NAME|false|false|USAGE|OnAccountObject|WAREHOUSE|WAREHOUSE_NAME" -> (known after apply)
        on_account        = false
        privileges        = [
            "USAGE",
        ]
        with_grant_option = false

      ~ on_account_object {
          ~ object_name = "\"WAREHOUSE_NAME\"" -> "WAREHOUSE_NAME" # forces replacement
            object_type = "WAREHOUSE"
        }
    }
```
Notice that the provider regards these resources as having been changed because of the quoted identifiers.

We do not want the grants to be dropped and recreated because that would require downtime, which we absolutely cannot have.

Thus, we attempted to upgrade to 0.95.0 (which contains your fix for the quoted identifiers) but that failed with:

>  Error: no schema available for module.resources.module.warehouse[0].snowflake_grant_privileges_to_role.warehouse_usage_grant["ROLE_NAME"] while reading state;

which is because resource type `snowflake_grant_privileges_to_role` no longer exists in 0.95.0.

hence we're between a rock and a hard place. We can't use 0.92.0 because it would cause all of our resources to be dropped and recreated. We can't use 0.95 because it no longer contains the definition of resource types that are in the state file.

Do you have any advice about how to tackle this?

**Comment by sfc-gh-jcieslak** [comment]
Hey @jamiekt 👋 
The simplest but cumbersome approach that I can think of right now is to use [lifecycle meta-argument](https://developer.hashicorp.com/terraform/language/meta-arguments/lifecycle) to `prevent_changes` on the fields that may cause replacement (also maybe `prevent_destroy` would help additionally). That should enable you to upgrade to v0.93.0 then after you upgrade to v0.95.0, you'll be able to drop `prevent_changes` meta-attributes and see no changes planned. It's worth noting that I'm not an expert in using Terraform, but that should work, afaik. Let me know If that approach works for you; if not, maybe I'll be able to find another way.

**Comment by jamiekt** [comment]
Thx @sfc-gh-jcieslak , well look into doing that.

**Comment by sfc-gh-jcieslak** [comment]
Hey 👋 

> This can't work, since these lifecycle arguments just cause the plan to throw errors.

Then, in this case, `prevent_destroy` shouldn't be used (only `prevent_changes` that would prevent plan changes on id and name).

> Similarly, I think ignore_changes=true would introduce state drift, i.e. the state would be mismatched with the code.

Yes, but in 0.95.0, that's handled, and you can remove the `prevent_changes` blocks there (the plan shouldn't be produced there).

> Have any SF clients reported issues due to these extraneous quotes?

Yes, that's why we planned to refactor them for a long time, and only recently did we find time to do so. Unfortunately, as the refactoring was concluded recently, only the latest version (0.95.0) has all the changes regarding identifier rework ([documented here](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/docs/technical-documentation/identifiers_rework_design_decisions.md)).

> Do we know for sure whether or not this will create duplicate roles with quotes in the actual role name?

If you refer to `snowflake_account_role`, in 0.95.0, it doesn't matter if the `name` is equal to `name = "\"role_name\"` or `name = "role_name"` (it will be treated as the same name and won't create duplicates).

I hope I addressed all the concerns. Please, try to proceed with migration without the `prevent_destroy` flag as suggested. If that fails, we'll consider other alternatives.

**Comment by jamiekt** [comment]
> If you refer to snowflake_account_role, in 0.95.0, it doesn't matter if the name is equal to name = "\"role_name\" or name = "role_name" (it will be treated as the same name and won't create duplicates).

That is very good news, thank you @sfc-gh-jcieslak

**Comment by sfc-gh-jcieslak** [comment]
@jamiekt Just to show how it behaves I did a simple test
```terraform
resource "snowflake_account_role" "test" {
  name = "TEST_ROLE_ONE" # results in TEST_ROLE_ONE (in Snowflake)
}

resource "snowflake_account_role" "test2" {
  name = "\"TEST_ROLE_TWO\"" # results in TEST_ROLE_TWO (in Snowflake)
}
```

Adding a third role with role name quoted internally would cause a naming clash with the one that is not quoted resulting in SQL compilation error: `Object 'TEST_ROLE_ONE' already exists.`.
```terraform
resource "snowflake_account_role" "test" {
  name = "TEST_ROLE_ONE"
}

resource "snowflake_account_role" "test2" {
  name = "\"TEST_ROLE_TWO\""
}

resource "snowflake_account_role" "test3" {
  name = "\"TEST_ROLE_ONE\"" # this is the same as the first one
}
```
Hope that helps.

**Comment by sfc-gh-jcieslak** [comment]
Closing due to long inactivity.

**Comment by jamiekt** [comment]
Hi @sfc-gh-jcieslak ,
I see you've closed this issue due to no activity from us which is fair enough. I'm happy to report that @kadu-vido and I have figured out an upgrade path that allows us to drop use of resource types that were removed in 0.93.0 and also circumvent this bug regarding embedded quotes. 
We're about to inform our user base of the work they're going to be required to do. As kadu said above we have >30 teams that need to do this work so its going to be a long process, but as least we now have a way forward.

**Comment by sfc-gh-jcieslak** [comment]
Hey @jamiekt,
I'm glad you and @kadu-vido managed to come up with the migration path, and if possible, we'd love to see this process to provide better support for the migration. If you've recently read our new post on the [roadmap](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/ROADMAP.md) we've managed to allocate some of our capacity to create grant migration helpers. We are planning to create a small tool(s) that will speed up the migration process for grant resources. The mentioned process could help in identifying the most difficult parts that we could speed up with the helpers. If you are interested, you can use your point of contact at Snowflake to share more information there.

**Comment by jamiekt** [comment]
Hi @sfc-gh-jcieslak ,
I'll share what I can. Here's the announcement we put out on our internal Slack, I've redacted some of the links so we don't expose company information

> [ourorg/snowflake-automation](https://github.com/redactedorg/snowflake-automation) is where we create and manage Snowflake databases, roles, users and warehouses. We do so using the [Snowflake provider for terraform](https://registry.terraform.io/providers/Snowflake-Labs/snowflake/latest/docs) for which Snowflake regularly release new versions. At the time of writing [most of our teams are using v0.92.0 of that provider](https://github.com/search?q=repo%3Aredactedorg%2Fsnowflake-automation%200.92.0&type=code) which was [released on 6th June 2024](https://github.com/Snowflake-Labs/terraform-provider-snowflake/releases/tag/v0.92.0).
v0.92.0 contained [a deprecation notice for 28 resource types](https://registry.terraform.io/providers/Snowflake-Labs/snowflake/0.92.0/docs#currently-deprecated-resources), many of which we use extensively in our organization, specifically in our [`resource_group_refactored` module](https://github.com/redactedorg/snowflake-automation/tree/master/terraform/modules/resource_group_refactored). These 28 resources were removed in v0.93.0 which is why so many of our teams are stuck on v0.92.0. Even though Snowflake provided a detailed [v0.92.0 ➞ v0.93.0 migration guide](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/MIGRATION_GUIDE.md#v0920--v0930) this has been described by one (slightly hyperbolic) observer as [the most hostile change I have ever seen on a project towards users](https://github.com/Snowflake-Labs/terraform-provider-snowflake/discussions/2736#discussioncomment-10558709).
To complicate matters [a bug was introduced in v0.92.0](https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2982) affecting the `import {...} block` which we must use to migrate away from the 28 removed resource types. It was [fixed in v0.95.0](https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2982#issuecomment-2331052648) but Snowflake [declined](https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2982#issuecomment-2296211672) to produce a patch for v.0.92.0 containing the fix. This put us in rather a difficult spot, we want to upgrade past v0.92.0 so we can stop using the deprecated resources but the bug prevents us from doing so in an orderly manner.
In 2024 some folks teamed up to figure out a migration plan for all teams using `resource_group_refactored` , this work was done under [Snowflake grants resource migration for Provider v0.93.0](https://redactedorg.atlassian.net/browse/DSOP-7) and [DSOP-134 Continue snowflake-automation update](https://redactedorg.atlassian.net/browse/DSOP-134). Today we’re pleased to announce that that work is complete and teams are now able to upgrade beyond v0.92.0. Upgrading is a two-step process:
>
> 1. Upgrade to `resource_group_refactored` v4.0.0 which replaces all the deprecated resource types. To prevent terraform dropping and recreating all the real-world objects one must define lots of import blocks so to ease the burden we have produced a file [imports.tf](https://github.com/redactedorg/snowflake-automation/blob/a97f9257051a22c74a4b4183af3ceeeedb4b0334/terraform/onboarding/imports.tf) with all those import blocks that you can copy/paste into your team's TF directory. Refer to PR [#4841 Onboarding provider upgrade](https://github.com/redactedorg/snowflake-automation/pull/4841) in which this step was carried out for [terraform/onboarding](https://github.com/redactedorg/snowflake-automation/tree/master/terraform/onboarding).
>
> 1. Upgrade to `resource_group_refactored` v4.0.1 which allows you to upgrade past v.0.92.0 of the Snowflake provider for terraform. Refer to PR [#4847 onboarding provider upgrade (continued)](https://github.com/redactedorg/snowflake-automation/pull/4847) in which this step was carried out for [terraform/onboarding](https://github.com/redactedorg/snowflake-automation/tree/master/terraform/onboarding).
>
> If these two steps are carried out correctly you will have upgraded to the new resource types and to a later version of the provider than v0.92.0 without affecting any of your real-world resources (and that's really the most important thing here).
>
> This is a complicated upgrade which we’ve tried to make as pain-free as possible via the provided file [imports.tf](https://github.com/redactedorg/snowflake-automation/blob/a97f9257051a22c74a4b4183af3ceeeedb4b0334/terraform/onboarding/imports.tf).
>
> Performing this upgrade is not mandatory however new capabilities will arrive in later versions of the provider that you may want to use (e.g. [v0.97.0 has a new ](https://registry.terraform.io/providers/Snowflake-Labs/snowflake/0.97.0/docs/resources/service_user)snowflake_service_user[ resource type](https://registry.terraform.io/providers/Snowflake-Labs/snowflake/0.97.0/docs/resources/service_user) that will be very useful) so at some point your own needs may require you to perform this upgrade. We recommend scheduling this work at the earliest opportunity.

Let me know if this helps and if there is any other help I can provide.

Apologies for taking a while to respond. I don't receive email notifications for this thread and I haven't figured out how to turn them on without turning the repo a spambot. Notification configuration on GitHub is a minefield. I'll remind myself to check back here every so often or I can be [contacted on linkedin](https://www.linkedin.com/in/jamiekthomson/).

>  If you are interested, you can use your point of contact at Snowflake to share more information there.

Yes please do that (I'm not sure who out point of contact is but I'm guessing you can look on linkedin to see who I work for there and then find them from that)

**Comment by sfc-gh-jcieslak** [comment]
Hey @jamiekt 👋 
You can send an email to swathi.thyagaraj@snowflake.com
and cc: jan.cieslak@snowflake.com, jakub.michalak@snowflake.com, artur.sawicki@snowflake.com.
